### PR TITLE
[WIP] Python 3 tests test (pretend you never saw it)

### PR DESCRIPTION
**Pull request by jrobeson**
This is a very forceful way to get the tests to pass on python3.
It needs to be totally redone in a python 2 compatible way.

### CODE REVIEW

**Comment by jrobeson** [comment]
sadly it broke more windows tests than I would have liked :(

File: beets/importer.py [hunk 1]
```diff
@@ -943,7 +943,9 @@ def is_archive(cls, path):
         if not os.path.isfile(path):
             return False
 
-        for path_test, _ in cls.handlers():
+        for path_test, x in cls.handlers():
+            if isinstance(path, bytes):
+                path = util.py3_path(path)
```
Comment (jrobeson) [inline beets/importer.py]: i assume we're seeing bytes here because the results from TarFile and friends are strings

File: beets/util/__init__.py [hunk 2]
```diff
@@ -216,8 +216,7 @@ def mkdirall(path):
             try:
                 os.mkdir(syspath(ancestor))
             except (OSError, IOError) as exc:
-                raise FilesystemError(exc, 'create', (ancestor,),
-                                      traceback.format_exc())
```
Comment (jrobeson) [inline beets/util/__init__.py]: This doesn't break in 3.5 strangely, but does in 3.4!

File: beets/ui/__init__.py [hunk 3]
```diff
@@ -74,6 +74,14 @@ class UserError(Exception):
...
 # Encoding utilities.
 
+def _convert_args(args):
+    if not args:
+        return None
+    for i, elem in enumerate(args):
+        if isinstance(elem, bytes):
+            args[i] = elem.decode(_arg_encoding())
+    return args
+
```
Comment (jrobeson) [inline beets/ui/__init__.py]: this obviously doesn't work quite as nicely in windows!

File: beets/importer.py [hunk 4]
```diff
@@ -990,7 +992,7 @@ def extract(self):
 
         try:
             extract_to = mkdtemp()
-            archive = handler_class(self.toppath, mode='r')
+            archive = handler_class(util.py3_path(self.toppath), mode='r')
```
Comment (sampsyo) [inline beets/importer.py]: This looks right!

File: beets/importer.py (hunk 1 above)
Comment (sampsyo) [inline beets/importer.py]: Can we test that assumption? If it's true, maybe we should just convert the results from those Unicode-only APIs back to our uniform representation (bytes).

File: beets/ui/__init__.py [hunk 5]
```diff
@@ -1034,7 +1042,7 @@ def parse_global_options(self, args):
         """Parse options up to the subcommand argument. Returns a tuple
         of the options object and the remaining arguments.
         """
-        options, subargs = self.parse_args(args)
+        options, subargs = self.parse_args(_convert_args(args))
```
Comment (sampsyo) [inline beets/ui/__init__.py]: Can you comment a little bit about why these calls are necessary?

File: beets/util/__init__.py (hunk 2 above)
Comment (sampsyo) [inline beets/util/__init__.py]: Huh! How does it break? s `format_exc` missing?

File: beets/util/__init__.py [hunk 6]
```diff
@@ -731,6 +726,12 @@ def command_output(cmd, shell=False):
     This replaces `subprocess.check_output` which can have problems if lots of
     output is sent to stderr.
     """
+    # Couldn't use _convert_args here due to circular dep of beets.ui
+    for i, elem in enumerate(cmd):
+        if isinstance(elem, six.text_type):
+            import locale
+            encoding = locale.getdefaultlocale()[1] or 'utf8'
+            cmd[i] = elem.encode(encoding)
```
Comment (sampsyo) [inline beets/util/__init__.py]: I think the right solution here is probably to track down all the places where Unicode is passed to `command_output` and fix those. We could add an `assert isinstance(arg, bytes)` for each argument to the top here; that would help find the roots of the problem.

Except, of course, on Windows—where Unicode arguments should probably be tolerated.

File: beetsplug/bpd/__init__.py [hunk 7]
```diff
@@ -640,8 +640,8 @@ class Command(object):
     """A command issued by the client for processing by the server.
     """
 
-    command_re = re.compile(br'^([^ \t]+)[ \t]*')
-    arg_re = re.compile(br'"((?:\\"|[^"])+)"|([^ \t"]+)')
+    command_re = re.compile(r'^([^ \t]+)[ \t]*')
+    arg_re = re.compile(r'"((?:\\"|[^"])+)"|([^ \t"]+)')
```
Comment (sampsyo) [inline beetsplug/bpd/__init__.py]: Can we instead keep these as ordinary bytes literals and double-escape the backslashes (i.e., replace every `\` with `\\`)?

File: beetsplug/convert.py [hunk 8]
```diff
@@ -183,11 +183,11 @@ def encode(self, command, source, dest, pretend=False):
...
         # Substitute $source and $dest in the argument list.
-        args = shlex.split(command)
+        args = shlex.split(command.decode(ui._arg_encoding()))
         for i, arg in enumerate(args):
             args[i] = Template(arg).safe_substitute({
-                'source': source,
-                'dest': dest,
+                'source': source.decode(ui._arg_encoding()),
+                'dest': dest.decode(ui._arg_encoding()),
             })
```
Comment (sampsyo) [inline beetsplug/convert.py]: Is the problem here that `Template` doesn't support bytes on Python 3?

File: beetsplug/hook.py [hunk 9]
```diff
@@ -58,7 +58,7 @@ def convert_field(self, value, conversion):
         converted = super(CodingFormatter, self).convert_field(value,
                                                                conversion)
         try:
-            converted = converted.decode(self._coding)
+            converted = converted.encode('utf-8').decode(self._coding)
```
Comment (sampsyo) [inline beetsplug/hook.py]: I don't quite see what's going on here—shouldn't `format_string` and `converted` be bytestrings here?

File: beetsplug/play.py [hunk 10]
```diff
@@ -138,7 +138,7 @@ def play_music(self, lib, opts, args):
             open_args = [self._create_tmp_playlist(paths)]
 
         self._log.debug(u'executing command: {} {}', command_str,
-                        util.text_string(b' '.join(open_args)))
+                        ui.decargs(open_args))
```
Comment (sampsyo) [inline beetsplug/play.py]: Since this is just a debug log, let's use `repr`. That is, we can use `{!r}` in the format string to get Unicode representations of arbitrary values.

File: test/_common.py [hunk 11]
```diff
@@ -53,7 +53,7 @@
 _item_ident = 0
 
 # OS feature test.
-HAVE_SYMLINK = hasattr(os, 'symlink')
+HAVE_SYMLINK = sys.platform != 'win32'
```
Comment (sampsyo) [inline test/_common.py]: As discussed in #2114, this is probably the right way to do the tests.

File: test/helper.py [hunk 12]
```diff
@@ -124,8 +124,8 @@ def has_program(cmd, args=['--version']):
     """
     full_cmd = [cmd] + args
     for i, elem in enumerate(full_cmd):
-        if isinstance(elem, six.text_type):
-            full_cmd[i] = elem.encode(_arg_encoding())
+        if isinstance(elem, bytes):
+            full_cmd[i] = elem.decode(_arg_encoding())
```
Comment (sampsyo) [inline test/helper.py]: This converts everything to Unicode strings. Is that necessary on Python 3? If so, should we leave an `if PY2:` here to do the opposite on Python 2?

File: test/test_importer.py [hunk 13]
```diff
@@ -414,7 +414,7 @@ class ImportTarTest(ImportZipTest):
     def create_archive(self):
         (handle, path) = mkstemp(dir=self.temp_dir)
         os.close(handle)
-        archive = TarFile(path, mode='w')
+        archive = TarFile(py3_path(path), mode='w')
```
Comment (sampsyo) [inline test/test_importer.py]: These uses of `py3_path` look right to me. Are there any lingering questions here?

File: test/test_template.py [hunk 14]
```diff
@@ -213,6 +213,7 @@ def test_nested_call_with_argument(self):
         self._assert_call(arg_parts[0], u"bar", 1)
         self.assertEqual(list(_normexpr(arg_parts[0].args[0])), [u'baz'])
 
+    @unittest.skip('use different data?')
```
Comment (sampsyo) [inline test/test_template.py]: Can you elaborate a little more about what's going wrong here?

File: test/test_types_plugin.py [hunk 15]
```diff
@@ -76,7 +76,7 @@ def test_float_modify_and_query(self):
 
         # Match in range
         out = self.list(u'myfloat:-10..0')
-        self.assertIn(b'aaa', out)
+        self.assertIn('aaa', out)
```
Comment (sampsyo) [inline test/test_types_plugin.py]: I suppose these changes mean we need to decide whether the captured `stdout` stream in the tests should be Unicode, bytes, or "native" (i.e., Unicode on 3, bytes on 2). I don't have a strong feeling about what we choose, but each has different implications for how these checks go.

File: beets/importer.py (hunk 1 above)
Comment (jrobeson) [inline beets/importer.py]: I definitely hedged on that  because I couldn't use pdb for these tests, and logging statements break the tests!.  I was hoping you could tell me.

It seems to first occur in the pretend import tests.  I can't see where I'm missing anything. I can tell you that it's track_2.mp3.

File: beets/ui/__init__.py (hunk 5 above)
Comment (jrobeson) [inline beets/ui/__init__.py]: because optparse only accepts strings, so it first breaks when seeing `--foo` type options

File: beets/util/__init__.py (hunk 2 above)
Comment (jrobeson) [inline beets/util/__init__.py]:
```sh
Traceback (most recent call last):
  File "/home/johnny/src/beets/test/test_files.py", line 508, in test_unsuccessful_move
    util.move(self.path, self.otherpath)
  File "/home/johnny/src/beets/beets/util/__init__.py", line 447, in move
    traceback.format_exc())
  File "/home/johnny/.local/share/pyenv/versions/3.4.5/lib/python3.4/traceback.py", line 256, in format_exc
    return "".join(format_exception(*sys.exc_info(), limit=limit, chain=chain))
  File "/home/johnny/.local/share/pyenv/versions/3.4.5/lib/python3.4/traceback.py", line 181, in format_exception
    return list(_format_exception_iter(etype, value, tb, limit, chain))
  File "/home/johnny/.local/share/pyenv/versions/3.4.5/lib/python3.4/traceback.py", line 146, in _format_exception_iter
    for value, tb in values:
  File "/home/johnny/.local/share/pyenv/versions/3.4.5/lib/python3.4/traceback.py", line 125, in _iter_chain
    context = exc.__context__
AttributeError: 'NoneType' object has no attribute '__context__'
```

File: beetsplug/bpd/__init__.py (hunk 7 above)
Comment (jrobeson) [inline beetsplug/bpd/__init__.py]: how would that help? the commands and args we're regexing against are strings

File: beetsplug/convert.py (hunk 8 above)
Comment (jrobeson) [inline beetsplug/convert.py]: yes, and neither does shlex.split

File: beetsplug/hook.py (hunk 9 above)
Comment (jrobeson) [inline beetsplug/hook.py]: it doesn't accept bytestrings, otherwise I wouldn't have had to change anything

File: beetsplug/play.py (hunk 10 above)
Comment (jrobeson) [inline beetsplug/play.py]: will do

File: test/_common.py (hunk 11 above)
Comment (jrobeson) [inline test/_common.py]: yeah, I didn't commit it because i though ti was a bad idea only to do it halfway and not also catch OSError like suggested

File: test/helper.py (hunk 12 above)
Comment (jrobeson) [inline test/helper.py]: if this is the approach we wanna take then yes, but I wondered if you had any better ideas.

File: test/test_importer.py (hunk 13 above)
Comment (jrobeson) [inline test/test_importer.py]: just that they were part of the overall issues with archive files, so i wanted to commit that as a whole change

File: test/test_template.py (hunk 14 above)
Comment (jrobeson) [inline test/test_template.py]: it doesn't raise a UnicodeDecodeError like it's supposed to. When I rebase again, I'll update the comment

File: test/test_types_plugin.py (hunk 15 above)
Comment (jrobeson) [inline test/test_types_plugin.py]: Please have a strong feeling :) hehe.  If we continue to use something that works like StringIO, then it's gonna be native strings.

File: beets/util/__init__.py (hunk 2 above)
Comment (jrobeson) [inline beets/util/__init__.py]: BTW: you wrote this comment in `HumanReadableException`

>    If `tb` is provided, it is a string containing a traceback for the
    associated exception. (Note that this is not necessary in Python 3.x
    and should be removed when we make the transition.)

File: beets/importer.py (hunk 1 above)
Comment (jrobeson) [inline beets/importer.py]: and now I can't replicate it anymore, I must have accidentally fixed it while doing other bytestring stuff (unless it comes back to bite me later)

File: beets/ui/__init__.py (hunk 5 above)
Comment (jrobeson) [inline beets/ui/__init__.py]: that is specifically it does `'--' + ouropt`  which is rarely ever a string, especially in our tests

File: beetsplug/bpd/__init__.py (hunk 7 above)
Comment (jrobeson) [inline beetsplug/bpd/__init__.py]: I realize I just don't get what you're trying to say here.  I'll try to clarify more specifically with an example from the bpd tests

```python
    def test_no_args(self):
        s = r'command'
        c = bpd.Command(s)
        self.assertEqual(c.name, u'command')
        self.assertEqual(c.args, [])
```

File: beets/importer.py (hunk 1 above)
Comment (sampsyo) [inline beets/importer.py]: Huh! OK, let's ignore this for now in that case.

File: beets/util/__init__.py (hunk 2 above)
Comment (sampsyo) [inline beets/util/__init__.py]: Yes, that's a good point—on Python 3, we should just use exception chaining (i.e., [PEP 3134](https://www.python.org/dev/peps/pep-3134/)) with a `from` clause on the `raise` statement. Maybe we can sidestep this issue by using `from` on Python 3 and `format_exc` on Python 2.

I still don't quite understand why that crashes, though! Quite mysterious indeed.

File: beetsplug/bpd/__init__.py (hunk 7 above)
Comment (sampsyo) [inline beetsplug/bpd/__init__.py]: Got it; thanks for clarifying—I assumed the commands were bytes.

It looks at first glance to me like the commands are Unicode on Python 2 also, right? If so, perhaps these changes are correct as-is?

File: beetsplug/convert.py (hunk 8 above)
Comment (sampsyo) [inline beetsplug/convert.py]: `shlex.split` doesn't work with bytes?! That's truly a crime! :angry:

Perhaps we need to rewrite the templating pieces with `.replace('$source', source)`. Otherwise, I can't see a way to exactly preserve the bytes in the filename when passing them to the command.

File: beetsplug/bpd/__init__.py (hunk 7 above)
Comment (jrobeson) [inline beetsplug/bpd/__init__.py]: I think you're right. I'll give it a go. There were so many little pieces just ot make it work, that I didn't get a chance to go over fix to see why things are the way they are.

File: beets/util/__init__.py (hunk 2 above)
Comment (jrobeson) [inline beets/util/__init__.py]: It's likely a bug, since it works fine on 3.5

File: beets/importer.py (hunk 1 above)
Comment (jrobeson) [inline beets/importer.py]: i already committed it without the bytes check, and everything so far is in order.

File: test/_common.py (hunk 11 above)
Comment (jrobeson) [inline test/_common.py]: I'm comitting it now because I think we're just missing an appropriate error message, but otherwise the functionality is ok

File: beetsplug/convert.py (hunk 8 above)
Comment (jrobeson) [inline beetsplug/convert.py]: Maybe I'm misunderstanding, but aren't they already not bytes on py3?

Comment (sampsyo) [inline beetsplug/convert.py]: `source` and `dest` are paths, so they're bytes on all versions. (At least on Unix, where paths are bytes "at heart.")

Comment (jrobeson) [inline beetsplug/convert.py]: but they aren't on py3, that's what all that convert_args business is for.

Comment (jrobeson) [inline beetsplug/convert.py]: this entire commit is just one way I've attempted to solve the problem of course, and I'm not wedded to it in any way shape or form. That's why  I wondered about us attempting to try using BytesIO and trying to see if maybe we could replace the stdin/stdout/stderr with bytes equivalents. I'm having a really hard time trying to find information on this in a way that works in both windows and linux. I read  a bunch of posts on this, but it's all either outdated or doesn't take into account both linux and windows. It also doesn't help us with stuff like shlex.split anyways

Comment (jrobeson) [inline beetsplug/convert.py]: Alternatively, I just have no idea what i'm talking about or doing, that's certainly a valid thought too in this case :)

**Comment by jrobeson** [comment]
I've rebased this again with all the decided stuff already pushed to master.

**Comment by jrobeson** [comment]
Almost everything left is hinging on us committing not dealing with args as bytes. If we intend to migrate to python 3 and also use click, then this is were we're gonna have to go in any case isn't it?

File: beetsplug/convert.py (hunk 8 above)
Comment (jrobeson) [inline beetsplug/convert.py]: I discovered that if you instantiate shlex, you can pass a file argument, thus in that case it could support bytes if a BytesIO or similar was passed to it.

File: beets/ui/__init__.py (hunk 3 above)
Comment (jrobeson) [inline beets/ui/__init__.py]: After reexamining I'm pretty sure this is really only needed for the ui tests that directly call `ui.raw_main` vs use `run_command`

File: beets/ui/__init__.py (hunk 5 above)
Comment (jrobeson) [inline beets/ui/__init__.py]: as mentioned above, this seems to be only necessary in the tests.

File: beets/util/__init__.py (hunk 2 above)
Comment (jrobeson) [inline beets/util/__init__.py]: this turned out to be waaay simpler than I thought.. seriously.  You're gonna laugh probably!.  Take a look here: https://github.com/beetbox/beets/blob/18b57ea02260039e76493b2baa4b1a3cefb17f15/beets/util/__init__.py#L446  and you'll see that there's no traceback for us to format. Python 2.x (and seemingly 3.5+)  are just letting us format non-existent tracebacks!

File: beets/ui/__init__.py (hunk 5 above)
Comment (jrobeson) [inline beets/ui/__init__.py]: as in, tests that call `ui._raw_main` directly.

**Comment by jrobeson** [comment]
What do you think about refactoring `run_command` in the test helper? ATM it implicitly uses `self.lib` if exists, or sets the library to `:memory`.    I'd like to be able to use `run_command` instead of calling `ui._raw_main`, but that breaks the tests due how `_raw_main` behaves when `lib` isn't passed to it (defaulting to `None`).

Any suggestions? Fixing that one way or the other would clean up this PR pretty nicely.

File: test/test_ui.py [hunk 16]
```diff
@@ -641,8 +641,8 @@ def setUp(self):
         self.io.install()
 
     def test_manual_search_gets_unicode(self):
-        self.io.addinput(b'\xc3\x82me')
-        self.io.addinput(b'\xc3\x82me')
+        self.io.addinput('Âme')
+        self.io.addinput('Âme')
```
Comment (jrobeson) [inline test/test_ui.py]: this one actually passes in 2 and 3.

**Comment by jrobeson** [comment]
@sampsyo: if you could look at these 2 tests in test_ui, then we could unmask py 3.4, and 3.5 from allowed failures.

File: test/test_ui.py (hunk 16 above)
Comment (sampsyo) [inline test/test_ui.py]: Seems like a fine strategy—we need a "native string" that's non-ASCII.

**Comment by sampsyo** [comment]
Happy to! I just pushed these two commits, which are only slight variants on what you already had:

https://github.com/beetbox/beets/commit/6772b4d1fd979d27c5d02fd2ef146dbccb8b0a9e
https://github.com/beetbox/beets/commit/0e2b8a5b60abbd7930c60627f536bf9f11cc1883

Do those look good to you?

**Comment by jrobeson** [comment]
yeah, I wasn't sure if you wanted to use `six`  stuff there, but since you do, then it's all good!

**Comment by jrobeson** [comment]
looks like we're done here

**Comment by sampsyo** [comment]
:tada: :x: :100: :exclamation:

**Awesome** work on this over the last few weeks. This is a huge milestone.
